src/storage/DataAccessorBasedStore.ts

```diff
@@ -0,0 +1,366 @@
+import { ConflictHttpError } from '../util/errors/ConflictHttpError';
+import { MethodNotAllowedHttpError } from '../util/errors/MethodNotAllowedHttpError';
+import { NotFoundHttpError } from '../util/errors/NotFoundHttpError';
+import { UnsupportedHttpError } from '../util/errors/UnsupportedHttpError';
+import type { MetadataController } from '../util/MetadataController';
+import { CONTENT_TYPE, HTTP, LDP, RDF } from '../util/UriConstants';
+import { ensureTrailingSlash, trimTrailingSlashes } from '../util/Util';
+import type { DataAccessor } from './accessors/DataAccessor';
+import type { ContainerManager } from './ContainerManager';
+import type { ResourceStore } from './ResourceStore';
+
+/**
+ * ResourceStore which uses a DataAccessor for backend access.
+ *
+ * The DataAccessor interface provides elementary store operations such as read and write.
+ * This DataAccessorBasedStore uses those elementary store operations
+ * to implement the more high-level ResourceStore contact, abstracting all common functionality
+ * such that new stores can be added by implementing the more simple DataAccessor contract.
+ * DataAccessorBasedStore thereby provides behaviours for reuse across different stores, such as:
+ *  * Converting container metadata to data
+ *  * Converting slug to URI
+ *  * Checking if addResource target is a container
+ *  * Checking if no containment triples are written to a container
+ *  * etc.
+ *
+ * Currently "metadata" is seen as something that is not directly accessible.
+ * That means that a user can't write directly to the metadata of a resource, only indirectly through headers.
+ * (Except for containers where data and metadata overlap).
+ *
+ * The one thing this store does not take care of (yet?) are containment triples for containers
+ *
+ * Work has been done to minimize the amount of required calls to the DataAccessor,
+ * but the main disadvantage is that sometimes multiple calls are required where a specific store might only need one.
+ */
+export class DataAccessorBasedStore implements ResourceStore {
+  private readonly accessor: DataAccessor;
+  private readonly base: string;
+  private readonly metadataController: MetadataController;
+  private readonly containerManager: ContainerManager;
+
+  public constructor(accessor: DataAccessor, base: string, metadataController: MetadataController,
+    containerManager: ContainerManager) {
+    this.accessor = accessor;
+    this.base = ensureTrailingSlash(base);
+    this.metadataController = metadataController;
+    this.containerManager = containerManager;
+  }
+
+  public async getRepresentation(identifier: ResourceIdentifier): Promise<Representation> {
+    this.validateIdentifier(identifier);
+
+    // In the future we want to use getNormalizedMetadata and redirect in case the identifier differs
+    const metadata = await this.accessor.getMetadata(identifier);
+
+    let result: Representation;
+    if (this.isExistingContainer(metadata)) {
+      // Container
```

Comment:
Suggestion of how I would approach this in the future. In my opinion, this delineates and describes individual blocks more clearly (external as opposed to internal comment, plus more explicit statement).

```suggestion
    // Create the representation of a container
    if (this.isExistingContainer(metadata)) {
```